File: solutions/2232-adding-spaces-to-a-string/adding-spaces-to-a-string.py
class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        i, result = 0, []

        for space in spaces:
            result.append(s[i : space])
            i = space

        result.append(s[i :])
        return " ".join(result)

File: solutions/2232-adding-spaces-to-a-string/adding-spaces-to-a-string.rb
def add_spaces(s, spaces)
    result = Array.new(s.length + spaces.length)
    write_pos = 0
    read_pos = 0
    
    spaces.each do |space_pos|
        while read_pos < space_pos
            result[write_pos] = s[read_pos]
            write_pos += 1
            read_pos += 1
        end
        result[write_pos] = ' '
        write_pos += 1
    end
    
    while read_pos < s.length
        result[write_pos] = s[read_pos]
        write_pos += 1
        read_pos += 1
    end
    
    result.join
end